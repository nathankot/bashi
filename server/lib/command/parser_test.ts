import { assertSnapshot } from "std/testing/snapshot.ts";
import {
  parseFunctionCall,
  parseFunctionCalls,
  parseActionGroup,
} from "./parser.ts";

for (const expr of [
  `someFunction(); someOtherFunction(" aa() ; bbb()")`,
  "someCall()",
  "someCall('a', 111)",
  "someCall('a \\'ha\\'', 111)",
  "someCall(`a`, 111)",
  "someCall(`a \\`ha\\``, 111)",
  `someCall("a", 111)`,
  `someCall("a \\"ha\\"", 111)`,
  `someCall("a", 111,true)`,
  `someCall("a", 111, false)`,
  `someCall("a false true 123 b", 111, false)`,
  `someCall(-123)`,
  `some_call()`,
  `some-call()`,
  `SOMECALL()`,
  `malformed("a"`,
  `someCall(-123.500)`,
  "testing blah",
  "",
]) {
  Deno.test(
    "parseFunctionCall(s): " + (expr === "" ? "empty string" : expr),
    (t) => {
      try {
        const result = parseFunctionCall(expr);
        assertSnapshot(t, result);
      } catch (e) {
        assertSnapshot(t, (e as Error).message);
      }

      try {
        const result = parseFunctionCalls(expr);
        assertSnapshot(t, result);
      } catch (e) {
        assertSnapshot(t, (e as Error).message);
      }
    }
  );
}

for (const expr of [
  `Thought: I need to do something
Action: someFunction(123, "str", true)
Result: blah blah blah blah`,
  `Thought: I need to do something
Action: someFunction(123, "str", true)
Result:`,
  `Thought: I need to do something
Action: someFunction(123, "str", true)
Result: `,
  `Thought: I need to do something action: thought: hmmm
Action: someFunction()  ; someOtherFunction(" aa() ; bbb()")
Result: blah blah blah blah`,
  `Thought: I need to do something
Action: someFunction(true); someOtherFunction(true, 123, 'str', "str2")`,
  `tHOUght: I need to do something
aCTion  :    someFunction(); someOtherFunction()`,
  `Thought: I need to do something Action: head fake

Action: someFunction(); someOtherFunction()
Result: blah blah blah blah`,

  // Invalid examples:

  `Thought no colon doesnt work\nAction hahaha`,
  `Action: action should not come first\nThought: ha`,
  `completely invalid`,
  ``,
]) {
  Deno.test(
    "parseActionGroup: " + (expr === "" ? "empty string" : expr),
    (t) => {
      try {
        const result = parseActionGroup(expr);
        assertSnapshot(t, result);
      } catch (e) {
        assertSnapshot(t, (e as Error).message);
      }
    }
  );
}
