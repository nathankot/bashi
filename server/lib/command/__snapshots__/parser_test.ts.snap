export const snapshot = {};

snapshot[`parseFunctionCall(s): someFunction(); someOtherFunction(" aa() ; bbb()") 1`] = `'{"index":14,"rowBegin":1,"columnBegin":15,"rowEnd":1,"columnEnd":16}: The parser cannot reach the end of file, stops at ";" at position {"index":14,"rowBegin":1,"columnBegin":15,"rowEnd":1,"columnEnd":16}.'`;

snapshot[`parseFunctionCall(s): someFunction(); someOtherFunction(" aa() ; bbb()") 2`] = `
[
  {
    args: [
    ],
    name: "someFunction",
    type: "call",
  },
  {
    args: [
      {
        type: "string",
        value: " aa() ; bbb()",
      },
    ],
    name: "someOtherFunction",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall() 1`] = `
{
  args: [
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall() 2`] = `
[
  {
    args: [
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall('a', 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a",
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall('a', 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a",
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall('a \\'ha\\'', 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a 'ha'",
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall('a \\'ha\\'', 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a 'ha'",
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(\`a\`, 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a",
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(\`a\`, 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a",
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(\`a \\\`ha\\\`\`, 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a \`ha\`",
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(\`a \\\`ha\\\`\`, 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a \`ha\`",
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall("a", 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a",
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall("a", 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a",
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall("a \\"ha\\"", 111) 1`] = `
{
  args: [
    {
      type: "string",
      value: 'a "ha"',
    },
    {
      type: "number",
      value: 111,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall("a \\"ha\\"", 111) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: 'a "ha"',
      },
      {
        type: "number",
        value: 111,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall("a", 111,true) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a",
    },
    {
      type: "number",
      value: 111,
    },
    {
      type: "boolean",
      value: true,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall("a", 111,true) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a",
      },
      {
        type: "number",
        value: 111,
      },
      {
        type: "boolean",
        value: true,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall("a", 111, false) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a",
    },
    {
      type: "number",
      value: 111,
    },
    {
      type: "boolean",
      value: false,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall("a", 111, false) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a",
      },
      {
        type: "number",
        value: 111,
      },
      {
        type: "boolean",
        value: false,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall("a false true 123 b", 111, false) 1`] = `
{
  args: [
    {
      type: "string",
      value: "a false true 123 b",
    },
    {
      type: "number",
      value: 111,
    },
    {
      type: "boolean",
      value: false,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall("a false true 123 b", 111, false) 2`] = `
[
  {
    args: [
      {
        type: "string",
        value: "a false true 123 b",
      },
      {
        type: "number",
        value: 111,
      },
      {
        type: "boolean",
        value: false,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(-123) 1`] = `
{
  args: [
    {
      type: "number",
      value: -123,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(-123) 2`] = `
[
  {
    args: [
      {
        type: "number",
        value: -123,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): some_call() 1`] = `
{
  args: [
  ],
  name: "some_call",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): some_call() 2`] = `
[
  {
    args: [
    ],
    name: "some_call",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): some-call() 1`] = `
{
  args: [
  ],
  name: "some-call",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): some-call() 2`] = `
[
  {
    args: [
    ],
    name: "some-call",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): SOMECALL() 1`] = `
{
  args: [
  ],
  name: "SOMECALL",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): SOMECALL() 2`] = `
[
  {
    args: [
    ],
    name: "SOMECALL",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): malformed("a" 1`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;

snapshot[`parseFunctionCall(s): malformed("a" 2`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;

snapshot[`parseFunctionCall(s): someCall(-123.500) 1`] = `
{
  args: [
    {
      type: "number",
      value: -123.5,
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(-123.500) 2`] = `
[
  {
    args: [
      {
        type: "number",
        value: -123.5,
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(true, a(b(), 123), "hello") 1`] = `
{
  args: [
    {
      type: "boolean",
      value: true,
    },
    {
      args: [
        {
          args: [
          ],
          name: "b",
          type: "call",
        },
        {
          type: "number",
          value: 123,
        },
      ],
      name: "a",
      type: "call",
    },
    {
      type: "string",
      value: "hello",
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(true, a(b(), 123), "hello") 2`] = `
[
  {
    args: [
      {
        type: "boolean",
        value: true,
      },
      {
        args: [
          {
            args: [
            ],
            name: "b",
            type: "call",
          },
          {
            type: "number",
            value: 123,
          },
        ],
        name: "a",
        type: "call",
      },
      {
        type: "string",
        value: "hello",
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(true, a(b(), 123), "hello",) 1`] = `
{
  args: [
    {
      type: "boolean",
      value: true,
    },
    {
      args: [
        {
          args: [
          ],
          name: "b",
          type: "call",
        },
        {
          type: "number",
          value: 123,
        },
      ],
      name: "a",
      type: "call",
    },
    {
      type: "string",
      value: "hello",
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(true, a(b(), 123), "hello",) 2`] = `
[
  {
    args: [
      {
        type: "boolean",
        value: true,
      },
      {
        args: [
          {
            args: [
            ],
            name: "b",
            type: "call",
          },
          {
            type: "number",
            value: 123,
          },
        ],
        name: "a",
        type: "call",
      },
      {
        type: "string",
        value: "hello",
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(true, a((), 123), "hello",) 1`] = `'{"index":17,"rowBegin":1,"columnBegin":18,"rowEnd":1,"columnEnd":19}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a((), 123), "hello",) 2`] = `'{"index":17,"rowBegin":1,"columnBegin":18,"rowEnd":1,"columnEnd":19}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true, 123), "hello",) 1`] = `
{
  args: [
    {
      type: "boolean",
      value: true,
    },
    {
      args: [
        {
          type: "boolean",
          value: true,
        },
        {
          type: "number",
          value: 123,
        },
      ],
      name: "a",
      type: "call",
    },
    {
      type: "string",
      value: "hello",
    },
  ],
  name: "someCall",
  type: "call",
}
`;

snapshot[`parseFunctionCall(s): someCall(true, a(true, 123), "hello",) 2`] = `
[
  {
    args: [
      {
        type: "boolean",
        value: true,
      },
      {
        args: [
          {
            type: "boolean",
            value: true,
          },
          {
            type: "number",
            value: 123,
          },
        ],
        name: "a",
        type: "call",
      },
      {
        type: "string",
        value: "hello",
      },
    ],
    name: "someCall",
    type: "call",
  },
]
`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123), "hello",) 1`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123), "hello",) 2`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123()), "hello",) 1`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123()), "hello",) 2`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123()), "hello"(),) 1`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): someCall(true, a(true(), 123()), "hello"(),) 2`] = `'{"index":21,"rowBegin":1,"columnBegin":22,"rowEnd":1,"columnEnd":23}: Unable to consume token: ('`;

snapshot[`parseFunctionCall(s): testing blah 1`] = `'{"index":8,"rowBegin":1,"columnBegin":9,"rowEnd":1,"columnEnd":13}: Unable to consume token: blah'`;

snapshot[`parseFunctionCall(s): testing blah 2`] = `'{"index":8,"rowBegin":1,"columnBegin":9,"rowEnd":1,"columnEnd":13}: Unable to consume token: blah'`;

snapshot[`parseFunctionCall(s): empty string 1`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;

snapshot[`parseFunctionCall(s): empty string 2`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;

snapshot[`parseActionGroup: Thought: I need to do something
Action: someFunction(123, "str", true)
Result: blah blah blah blah 1`] = `
{
  action: 'someFunction(123, "str", true)',
  functionCalls: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "str",
        },
        {
          type: "boolean",
          value: true,
        },
      ],
      name: "someFunction",
      type: "call",
    },
  ],
  result: "blah blah blah blah",
  thought: "I need to do something",
}
`;

snapshot[`parseActionGroup: Thought: I need to do something
Action: someFunction(123, "str", true)
Result: 1`] = `
{
  action: 'someFunction(123, "str", true)',
  functionCalls: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "str",
        },
        {
          type: "boolean",
          value: true,
        },
      ],
      name: "someFunction",
      type: "call",
    },
  ],
  result: undefined,
  thought: "I need to do something",
}
`;

snapshot[`parseActionGroup: Thought: I need to do something
Action: someFunction(123, "str", true)
Result:  1`] = `
{
  action: 'someFunction(123, "str", true)',
  functionCalls: [
    {
      args: [
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "str",
        },
        {
          type: "boolean",
          value: true,
        },
      ],
      name: "someFunction",
      type: "call",
    },
  ],
  result: undefined,
  thought: "I need to do something",
}
`;

snapshot[`parseActionGroup: Thought: I need to do something action: thought: hmmm
Action: someFunction()  ; someOtherFunction(" aa() ; bbb()")
Result: blah blah blah blah 1`] = `
{
  action: 'someFunction()  ; someOtherFunction(" aa() ; bbb()")',
  functionCalls: [
    {
      args: [
      ],
      name: "someFunction",
      type: "call",
    },
    {
      args: [
        {
          type: "string",
          value: " aa() ; bbb()",
        },
      ],
      name: "someOtherFunction",
      type: "call",
    },
  ],
  result: "blah blah blah blah",
  thought: "I need to do something action: thought: hmmm",
}
`;

snapshot[`parseActionGroup: Thought: I need to do something
Action: someFunction(true); someOtherFunction(true, 123, 'str', "str2") 1`] = `
{
  action: \`someFunction(true); someOtherFunction(true, 123, 'str', "str2")\`,
  functionCalls: [
    {
      args: [
        {
          type: "boolean",
          value: true,
        },
      ],
      name: "someFunction",
      type: "call",
    },
    {
      args: [
        {
          type: "boolean",
          value: true,
        },
        {
          type: "number",
          value: 123,
        },
        {
          type: "string",
          value: "str",
        },
        {
          type: "string",
          value: "str2",
        },
      ],
      name: "someOtherFunction",
      type: "call",
    },
  ],
  result: undefined,
  thought: "I need to do something",
}
`;

snapshot[`parseActionGroup: tHOUght: I need to do something
aCTion  :    someFunction();; someOtherFunction() 1`] = `
{
  action: "someFunction();; someOtherFunction()",
  functionCalls: [
    {
      args: [
      ],
      name: "someFunction",
      type: "call",
    },
    {
      args: [
      ],
      name: "someOtherFunction",
      type: "call",
    },
  ],
  result: undefined,
  thought: "I need to do something",
}
`;

snapshot[`parseActionGroup: Thought: I need to do something Action: head fake

Action: someFunction(); someOtherFunction()
Result: blah blah blah blah 1`] = `
{
  action: "someFunction(); someOtherFunction()",
  functionCalls: [
    {
      args: [
      ],
      name: "someFunction",
      type: "call",
    },
    {
      args: [
      ],
      name: "someOtherFunction",
      type: "call",
    },
  ],
  result: "blah blah blah blah",
  thought: "I need to do something Action: head fake",
}
`;

snapshot[`parseActionGroup: Thought no colon doesnt work
Action hahaha 1`] = `
'{"index":8,"rowBegin":1,"columnBegin":9,"rowEnd":1,"columnEnd":9}: Unable to tokenize the rest of the input: no colon doesnt work
Action hahaha'
`;

snapshot[`parseActionGroup: Action: action should not come first
Thought: ha 1`] = `'{"index":0,"rowBegin":1,"columnBegin":1,"rowEnd":1,"columnEnd":7}: Unable to consume token: Action'`;

snapshot[`parseActionGroup: completely invalid 1`] = `'{"index":0,"rowBegin":1,"columnBegin":1,"rowEnd":1,"columnEnd":1}: Unable to tokenize the rest of the input: completely invalid'`;

snapshot[`parseActionGroup: empty string 1`] = `"<END-OF-FILE>: Unable to consume token: <END-OF-FILE>"`;
